### ISSUE TITLE
JSRuntime for non web builds (JavaScriptBridge improvement)

### ISSUE THREAD

**Issue by telamon** [godotengine/godot-proposals#10748]
### Describe the project you are working on

I'm hacking on a game where the game-core is written in NodeJS/custom-web-runtime.
It's relying heavily on Godot's JavaScriptBridge for game-state and multiplayer networking.

### Describe the problem or limitation you are having in your project

JSBridge dosen't work unless the project is running in a web environment.

I cannot test the game through **F5** (Run Project) / **F6** (Run Current Scene), so i have to do a web-export every time i do a change.
I would also like to release the game as native/mobile, but would need an godot owned JS-runtime for that. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Optionally bundle a minimal JS-runtime so bridge works on mobile/desktop builds.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When starting Godot JavaScriptBridge,
if runtime is not available,
then boot node.js or a bundled [bare](https://github.com/holepunchto/bare/tree/main) thread.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I have abstracted/stubbed out most JSBridge calls with hardcoded return values but i walked past the warning signs of spaghetti 700 lines ago.

### Is there a reason why this should be core and not an add-on in the asset library?

I cannot answer this, please help?

**Comment by btarg** [comment]
This would improve the iteration time of web exports a lot!

**Comment by telamon** [comment]
@btarg no. not at the moment i'm afraid. In my case I understand why the core is written in JS and i suspect the choice is correct. ( The game depends on js package `hyperswarm` + other js code that is just quicker to iterate through unit-tests )

_Sidenote:_
I started piecing together an `godot-bare` extension but quickly realized that I was out of my depth,
does anyone know if it's even be possible to hook existing JSBridge into a custom prebooted runtime via cpp GDExtension?

**Comment by Calinou** [comment]
Like for WebViews, I really feel this should be an add-on as bundling a JavaScript runtime in core would increase binary size a fair bit and increase the attack surface for security vulnerabilities.

**Comment by telamon** [comment]
@Calinou yes a js runtime with node-api's, not browser. I understand it's a very powerful addition and i suppose the full weight of including such a plugin in your project would be ~50MB plus user-sources, i don't think it should be enabled by default.

<details>
<summary><i>my usecase</i></summary>

I've built a non-traditional multiplayer game that uses DHT/Swarm for connections and world state processing.  
I use Godot to drive the core logic and alter the world state, mostly because I enjoy working with godot.


Currently the only runtime that supports the full game features is pears/keet but as i mentioned previously, the wasm compilation +  [patching](https://github.com/telamon/pear-godot/blob/master/pearsgodot_index.js.patch) makes development painfully slow.
I could ditch godot and pears in favour of a terminal-ui and run vanilla node-js, but that's like the last way out.

If I could toggle the JSBridge to bridge to something like bare/bun/node/deno in standard builds, I'd be able to cut iteration time and release native builds for desktop and even mobile. Again I'm mentioning 'bare' because it's cross device compatible/embeddable and somewhat minimal.  
I'm afraid i can't accomplish this myself, suggestions appreciated.
</details>

**Comment by FlooferLand** [comment]
Could a web-view be used for this instead of packaging a runtime?

Windows 10 and 11 have web-views built into the OS via Edge, allowing apps to pull up a web browser _(for logins, etc)_ without including a web browser in their binary.
Mac most definitely supports it via Safari.
Linux supports it as well but it's a bit more tricky.

Some frameworks such as [Tauri](https://tauri.app/) implement web-views as a way to replace integrating browsers in apps _(Electron 🤮)_, and Tauri works on all desktop and mobile platforms.

**Comment by telamon** [comment]
@FlooferLand Um. thank you for the input.. But the system web-view does not have node-api's, 
I would like to plug Godot's Javascript Bridge into a headless runtime when godot is running natively (not in browser) instead of the bridge just being limp.

Like this proposal is about adding a [runtime](https://github.com/holepunchto/bare/) (backend) _not_ a [webview](https://github.com/webview/webview) (frontend). :/
